Adds `hasLink` test for examples; update ~37 descriptions to pass

domlet opened the pull request:
This PR adds a test to require that each GL JS example has at least one link URL in the markdown description.

It includes updates to ~37 examples so they pass the test.

domlet commented:
Any @mapbox/docs want to volunteer to eyeball these changes to GL JS descriptions before I tag GL JS team for accuracy/content review? Thank you!

HeyStenson commented on docs/pages/example/filter-features-within-map-view.md:
```diff
@@ -15,6 +15,6 @@ prependJs:
 ---
 
-Move the map to query viewable features in a vector tile layer and filter by typing in an input.
+Use [`queryRenderedFeatures`](/mapbox-gl-js/api/map/#map#queryrenderedfeatures) to enable a user to move a map to query viewable features in a vector tile layer and filter by typing in an input.
```
```suggestion
This example uses [`queryRenderedFeatures`](/mapbox-gl-js/api/map/#map#queryrenderedfeatures) to restrict a list of features in a vector tile layer to only return those that are visible in the map view. 
```

HeyStenson commented on docs/pages/example/interactive-false.md:
```diff
@@ -15,6 +15,6 @@ prependJs:
 ---
 
-Set `interactive: false` to create a static map.
+Set the map parameter [`interactive`](/mapbox-gl-js/api/map/#map-parameters) to `false` to present a static map with no interactivity.
```
```suggestion
Set the `Map` object's [`interactive`](/mapbox-gl-js/api/map/#map-parameters) parameter to `false` to create a static map with no interactivity.
```

HeyStenson commented on docs/pages/example/live-update-feature.md:
```diff
@@ -15,6 +15,8 @@ prependJs:
 ---
 
-Change an existing feature on your map in real-time by updating its data.
+This example changes an existing feature on a map in real-time by updating its data.
```
Someone on the GL JS team will be able to confirm, but I believe the original description here is a little hand-wavy. I _think_ it's more accurate to say that this example shows how to visualize the progression of a path by drawing coordinates to the map one at a time, which could be used to deliver real-time data?

domlet replied:
Thank you. Changing from this:

> This example changes an existing feature on a map in real-time by updating its data.
>
> It also uses [`panTo`](/mapbox-gl-js/api/map/#map#panto) to animate the camera as the location of the `line` feature changes.

to this in 183c912 and would love to know from GL JS folks if this is accurate.

> This example shows how to change an existing feature on a map by updating its data.
> 
> It shows the progression of a path by drawing new coordinates to a feature in a [`line`](https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/#line) type layer, one at a time, an approach which you can use with a real-time data source.
> 
> It also uses [`panTo`](/mapbox-gl-js/api/map/#map#panto) to animate the camera to follow the line as new coordinates are added.

ryanhamley replied:
>It shows the progression of a path by ~drawing~ adding new coordinate pairs to a feature in a line ~type~ layer, ~one at a time,~ an approach ~which~ you can use with a real-time data source.

Calling `setData` kicks off a new render cycle leading to the line updating in real-time without having to ever explicitly animate anything. `panTo` then follows the leading edge of the line to keep it on screen.

domlet replied:
Love it – some of these changes were on remote; here's an updated description based your your remarks here:

> This example shows how to change an existing feature on a map by updating its data.
> 
> It shows the progression of a path by adding new coordinates to a feature in a [`line` layer](/mapbox-gl-js/style-spec/layers/#line). This approach is useful for visualizing real-time data sources.
> 
> Calling [`setData`](/mapbox-gl-js/api/sources/#geojsonsource#setdata) begins a new render cycle which makes the updates appear in real time without explicitly creating an animation. [`panTo`](/mapbox-gl-js/api/map/#map#panto) then follows the leading edge of the line to keep it on screen.

HeyStenson reviewed (commented):
I still have a lot to look at but wanted you to be able to get started with the ones I've reviewed! One thing that stood out to me wording-wise was the pattern of starting the first paragraph with "This example..." and then the second paragraph with "This application..." or "The application...". It was a little jarring to me as the reader - I'd prefer one or the other, but not both.

HeyStenson commented on docs/pages/example/video-on-a-map.md:
```diff
@@ -16,6 +16,8 @@ prependJs:
 ---
 
-Display a video on top of a satellite raster baselayer. Click the map to play and pause the video.
+Display a [`video`](/mapbox-gl-js/style-spec/sources/#video) in a map layer by adding a [source](/mapbox-gl-js/style-spec/sources/) and then adding a [layer](/mapbox-gl-js/style-spec/layers/) using that source. This example adds a video layer on top of a raster layer with satellite.
+
+Add interactivity by using the [click](/mapbox-gl-js/api/map/#map.event:click) event to pause and play the video.
```
```suggestion
Display a [`video`](/mapbox-gl-js/style-spec/sources/#video) in a map layer by adding a [source](/mapbox-gl-js/style-spec/sources/) and then adding a [layer](/mapbox-gl-js/style-spec/layers/) using that source. This example adds a video layer on top of a raster layer with satellite imagery.

Add interactivity by using the [`click`](/mapbox-gl-js/api/map/#map.event:click) event to pause and play the video.
```

HeyStenson commented on docs/pages/example/updating-choropleth.md:
```diff
@@ -16,6 +16,12 @@ prependJs:
 ---
 
-Use 2014 census data to display state or county population depending on zoom level.
+This map uses 2014 U.S. Census data to create two layers, one containing a choropleth visualization for _state populations_, and one containing a choropleth visualization for _county populations_. 
+
+Then, it defines a zoom level at which each layer and its respective legend should appear or disappear.
+
+You can use the `minzoom` and `minzoom` [layer properties](/mapbox-gl-js/style-spec/layers/) to define the zoom levels at which a map layer should appear or disappear.
+
+You can use [`getZoom`](/mapbox-gl-js/api/map/#map#getzoom) with the [`MapBoxZoomEvent`](/mapbox-gl-js/api/events/#mapboxzoomevent) event type to get zoom values to control the appearance of other elements. 
```
```suggestion
This map uses 2014 U.S. Census data to create two layers. One layer contains a choropleth visualization for _state populations_, and one layer contains a choropleth visualization for _county populations_. 

The example defines a zoom level at which each layer and its respective legend should appear or disappear using the `minzoom` and `minzoom` [layer properties](/mapbox-gl-js/style-spec/layers/).

You can use [`getZoom`](/mapbox-gl-js/api/map/#map#getzoom) with the [`MapBoxZoomEvent`](/mapbox-gl-js/api/events/#mapboxzoomevent) event type to get zoom values to control the appearance of other elements. 
```

HeyStenson commented on docs/pages/example/scroll-fly-to.md:
```diff
@@ -15,6 +15,8 @@ prependJs:
 ---
 
-Scroll down through the story and the map will fly to the chapter's location.
+This example shows a map on the left and a scrollable story with several chapters on the right. A user can scroll through the story and the map will fly to each chapter's location.
+
+When the user scrolls, the application recognizes which chapter is on screen, and uses [`flyTo`](/mapbox-gl-js/api/map/#map#flyto) to animate the transition to that chapter's unique `center` location and `bearing`, `zoom`, and `pitch` [`CameraOptions`](/mapbox-gl-js/api/properties/#cameraoptions) on the map.
```
```suggestion
This example shows a map on the left and a scrollable story with several chapters on the right. A user can scroll through the story and the map will fly to the corresponding location for each chapter.

When the user scrolls, the application recognizes which chapter is on screen, and uses [`flyTo`](/mapbox-gl-js/api/map/#map#flyto) to animate the transition to that chapter's unique `center` location and `bearing`, `zoom`, and `pitch` [`CameraOptions`](/mapbox-gl-js/api/properties/#cameraoptions) on the map.
```

HeyStenson commented on docs/pages/example/satellite-map.md:
```diff
@@ -16,6 +16,8 @@ prependJs:
 ---
 
-Display a satellite raster baselayer.
+This example shows how to display a satellite raster baselayer on a web map.
```
I think we should reword this. I think since this was originally written we've since started using "baselayer" to refer specifically to styles designed in Studio Classic (https://docs.mapbox.com/help/glossary/baselayer/). Maybe:
```suggestion
This example shows how to initialize a map using the [Mapbox Satellite](https://www.mapbox.com/maps/satellite/) style.
```

domlet replied:
I was wondering about this – thank you, I'll remove the other baselayer references as well.

HeyStenson commented on docs/pages/example/query-similar-features.md:
```diff
@@ -15,6 +15,10 @@ prependJs:
-Hover over counties to highlight counties that share the same name.
+In this example, a user can hover their cursor over any county in the United States, and the map highlights other U.S. counties with a matching name. The application also shows a popup with the name of the county at the cursor location, and shows an HTML overlay with summary information.
+
+The example first uses [addSource](/mapbox-gl-js/api/map/#map#addsource) to connect data from a vector tileset, and then uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add two [`fill`](/mapbox-gl-js/style-spec/layers/#fill) layers using that source.
```
```suggestion
The example first uses [`addSource`](/mapbox-gl-js/api/map/#map#addsource) to connect data from a vector tileset, and then uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add two [`fill`](/mapbox-gl-js/style-spec/layers/#fill) layers using that source.
```

HeyStenson commented on docs/pages/example/query-similar-features.md:
```diff
@@ -15,6 +15,10 @@ prependJs:
+
+The example first uses [addSource](/mapbox-gl-js/api/map/#map#addsource) to connect data from a vector tileset, and then uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add two [`fill`](/mapbox-gl-js/style-spec/layers/#fill) layers using that source.
+
+Then, it uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to gain [MapMouseEvent](/mapbox-gl-js/api/events/#mapmouseevent) information to use with [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to search for the county located at the cursor position, and other counties with the same name.
```
```suggestion
It uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to find the name of the county located at the cursor position, then uses and [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to find other counties with the same name and highlights them using an overlay on the map.
```

domlet commented on docs/pages/example/add-image-generated.md:
```diff
@@ -16,6 +16,6 @@ prependJs:
 ---
 
-Add an icon to the map that was generated at runtime.
+This example generates an icon at runtime, then uses [`addImage`](/mapbox-gl-js/api/map/#map#addimage) to add it to the map.
```
Thank you! Updating in a822732fa87ddd19bf272f1d927514c788db2dd1

domlet commented on docs/pages/example/disable-scroll-zoom.md:
```diff
@@ -15,6 +15,6 @@ prependJs:
 ---
 
-Disable scroll zoom to prevent scroll from zooming a map.
+This example uses `scrollZoom.disable` so the [`ScrollZoomHandler`](/mapbox-gl-js/api/handlers/#scrollzoomhandler) prevents a user's scroll interactions from zooming a map.
```
Thank you for your fine attention!

domlet commented on docs/pages/example/drag-a-point.md:
```diff
@@ -15,6 +15,8 @@ prependJs:
-Drag the point to a new location on a map and populate its coordinates in a display.
+In this example, a user can drag a point to a new location on the map, and a map overlay displays the new point coordinates.
+
+The application stores the point coordinates in a variable with a GeoJSON value, then uses [`MapTouchEvent`s](/mapbox-gl-js/api/events/#maptouchevent) to prevent the default map drag behavior and enable changes to the variable.
```
Done in f271351 – thank you!

HeyStenson commented on docs/pages/example/animate-a-line.md:
```diff
@@ -16,6 +16,10 @@ prependJs:
-Animate a line by updating a GeoJSON source on each frame.
+This example animates a line by updating a GeoJSON source on each frame.
+
+The application uses a variable to store a GeoJSON object, then uses [adds a source](/mapbox-gl-js/api/map/#map#addsource) referring to that object as a data value, then uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add a [line type layer](/mapbox-gl-js/style-spec/layers/#line) to the map.
```
```suggestion
The application uses a variable to store a GeoJSON object, then uses [`addSource`](/mapbox-gl-js/api/map/#map#addsource) referring to that object as a data value, then uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add a [`line` type layer](/mapbox-gl-js/style-spec/layers/#line) to the map.
```

HeyStenson commented on docs/pages/example/attribution-position.md:
```diff
@@ -15,6 +15,8 @@ prependJs:
 ---
 
-Place attribution in the `top-left` position when initializing a map.
+This example places the [required](https://docs.mapbox.com/help/how-mapbox-works/attribution/) map attribution in the `top-left` position when initializing a map.
+
+The application uses [`addControl`]https://docs.mapbox.com/mapbox-gl-js/api/map/#map#addcontrol with [`AttributionControl`](/mapbox-gl-js/api/markers/#attributioncontrol) to specify the position for this control as `top-left`.
```
```suggestion
This example places the [required map attribution](https://docs.mapbox.com/help/how-mapbox-works/attribution/) in the `top-left` position of a map using [`addControl`](/mapbox-gl-js/api/map/#map#addcontrol) with [`AttributionControl`](/mapbox-gl-js/api/markers/#attributioncontrol).
```

HeyStenson commented on docs/pages/example/dancing-buildings.md:
```diff
@@ -15,6 +15,10 @@ prependJs:
 ---
 
-Connect the runtime styling API with the Web Audio API to create a map where the 3D buildings dance to the rhythm of your ambient environment.
+This example connects the runtime styling API with the Web Audio API to create a map where the 3D buildings dance to the rhythm of your ambient environment.
```
I think we need to sub "expressions" for "runtime styling API" here, but someone on the GL JS team will know for sure:

 ```suggestion
This example uses the [Web Audio API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API) along with the power of [expressions](/mapbox-gl-js/style-spec/expressions/) to create a map where the 3D buildings dance to the rhythm of your ambient environment.
```

domlet replied:
I'll leave this open for comment from GL JS team. Thank you!

HeyStenson commented on docs/pages/example/query-similar-features.md:
```diff
@@ -15,6 +15,10 @@ prependJs:
+
+The example first uses [`addSource`](/mapbox-gl-js/api/map/#map#addsource) to connect data from a vector tileset, and then uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add two [`fill`](/mapbox-gl-js/style-spec/layers/#fill) layers using that source.
+
+It uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to find the name of the county located at the cursor position, then uses and [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to find other counties with the same name and highlights them using an overlay on the map.
```
```suggestion
It uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to find the name of the county located at the cursor position, then uses [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to find other counties with the same name and highlights them using an overlay on the map.
```

HeyStenson commented on docs/pages/example/cluster-html.md:
```diff
@@ -15,6 +15,8 @@ prependJs:
-This advanced example uses Mapbox GL JS clustering with HTML markers and custom property expressions. To use HTML or SVG for clusters in place of a Mapbox GL layer, you have to manually synchronize the clustered source with a pool of marker objects that updates continuously while the map view changes.
+This advanced example uses Mapbox GL JS [clustering](/mapbox-gl-js/style-spec/sources/#geojson-cluster) with HTML [markers](/mapbox-gl-js/api/markers/#marker) and the custom property [expressions](/mapbox-gl-js/style-spec/expressions/) `number-format`, `get`, `!=`, `>`, `>=`, `all`, and `case`. 
+
+You can use HTML or SVG for clusters in place of a Mapbox GL layer by manually synchronizing the clustered source with a pool of marker objects that updates continuously while the map view changes.
```
```suggestion
You can use HTML or SVG for clusters in place of a Mapbox GL JS layer by manually synchronizing the clustered source with a pool of marker objects that updates continuously while the map view changes.
```

❓

domlet replied:
I think the layer is Mapbox GL (not GL JS) because map layers exist in Mapbox beyond GL JS.

HeyStenson reviewed (commented):
This is great @domlet - what a huge undertaking! I've added some comments, and am happy to take another look after someone/s from the GL JS team review if that would be helpful to you 🎉

domlet commented:
> One thing that stood out to me wording-wise was the pattern of starting the first paragraph with "This example..." and then the second paragraph with "This application..." or "The application...". It was a little jarring to me as the reader - I'd prefer one or the other, but not both.

@HeyStenson I appreciate this. I've removed instances of paragraphs beginning with "This application" or "The application" but there are a couple instances of "application" inside of description paragraphs. 

I've updated style guide here: https://github.com/mapbox/documentation/pull/603 – increase strictness welcome.

I think I may add a test to force all descriptions in the body to begin with "This example..."

domlet commented on docs/pages/example/add-image-missing-generated.md:
```diff
@@ -16,6 +16,6 @@ prependJs:
 ---
 
-Dynamically generate a missing icon at runtime and add it to the map.
+This example dynamically generates a missing icon at runtime, then uses [`addImage`](/mapbox-gl-js/api/map/#map#addimage) to add it to the map.
```
Thank you! Leaving this open to ask for guidance from GL JS team 👀 

1. Does `addImage` add the image to a style, or a style layer? Asking re: these 2 descriptions:

> This example uses [`addImage`](/mapbox-gl-js/api/map/#map#addimage) to dynamically generate an icon at runtime and add it to a map style. Then, it adds the icon to the map by including the `'layout': {'icon-image': ['concat', 'square-rgb-', ['get', 'color']]}` statement inside [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer).

^ From add-image-missing-generated.md

> This example uses [`addImage`](/mapbox-gl-js/api/map/#map#addimage) to generate an icon at runtime and add it to a map style. Then, it adds the icon to the map by including the `'layout': {'icon-image': 'gradient'}` statement inside [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer).

^ From add-image-generated.md

2. Another question – I have never understood why we describe the icon here as 'missing'. Can we add a sentence to articulate under what circumstances the tool (I'm assuming Styles API) is understanding the icon to be missing?

Thank you!

ryanhamley replied:
1. Icons and images are per style and they're referenced by sources and layers. So it's accurate to say that the image is added to the style.

2. The most common way for images to be added to a style is through the sprite property. If a source or layer then references an image that isn't in the sprite, we'd say that image is missing and the developer should call `addImage` to add it to the style and make it available. But really any image referenced by a source or layer that's not in the style's list of available images can be said to be missing. In that case, the property that references the image would end up throwing the `styleimagemissing` event which would give the developer the chance to recognize that the image is missing and add it to the style.

I'm not sure the best way to re-word this, but I agree it's a little oddly worded. It sounds like you're generating a "missing icon", whatever that is, rather than "an icon that is not available in the style" if you know what I mean. The most technically correct way to phrase what's happening is that we're adding the `points` layer at runtime which references an `icon-image` that does not exist. This throws the `styleimagemissing` event which we're listening for and when that event happens, we generate an image and add it to the style with the image name referenced by the layer.

domlet replied:
Super super helpful. Here's the new description: 94b50f3

> This example uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add a layer to a [style](https://docs.mapbox.com/help/glossary/style/), but it references an image that does not exist in the style's [sprite](https://docs.mapbox.com/help/glossary/sprite) or list of available images. Since the image is missing, the property that references the image throws a [`styleimagemissing`](/mapbox-gl-js/api/map/#map.event:styleimagemissing) event.
> 
> This example uses that `styleimagemissing` event to detect the missing image, then it generates a new image and adds that image to the map style with [`addImage`](/mapbox-gl-js/api/map/#map#addimage) within the event listener callback to make sure that an icon is rendered even if the referenced image is missing.

domlet commented:
Friends in @mapbox/gl-js, would someone review this PR to help us ensure that ~37 expanded example descriptions are accurate?

* We currently have a total of 124 [GL JS examples](https://docs.mapbox.com/mapbox-gl-js/example/).
* This PR enhances ~37 descriptions by adding more detailed language and adding hyperlinks to docs, improving ~30% of our existing content in this section.

All GL JS examples now require at least one link in the example description in the body.

This enhancement improves our users' ability to understand the content of existing examples, and differentiate between docs content of the `example` type versus the `tutorial` type.

Thank you!

![image](https://user-images.githubusercontent.com/6026447/104474659-16f9c600-5573-11eb-9188-2a88899ca2cc.png)

ryanhamley commented on docs/pages/example/animate-a-line.md:
```diff
@@ -16,6 +16,10 @@ prependJs:
-Animate a line by updating a GeoJSON source on each frame.
+This example animates a line by updating a GeoJSON source on each frame.
+
+It uses a variable to store a GeoJSON object, then uses [`addSource`](/mapbox-gl-js/api/map/#map#addsource) referring to that object as a data value, then uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add a [`line` type layer](/mapbox-gl-js/style-spec/layers/#line) to the map.
```
I don't think you need to call out saving the GeoJSON into a variable. "This example adds a GeoJSON source and line layer to the map. When the animation begins, points are calculated and added to the source with `setData`."

At a basic level, this is pretty much the same approach as https://docs.mapbox.com/mapbox-gl-js/example/live-update-feature/

domlet replied:
Fabulous, thank you. 2e1a1ca

ryanhamley commented on docs/pages/example/add-image-missing-generated.md:
```diff
@@ -16,6 +16,6 @@ prependJs:
 ---
 
-Dynamically generate a missing icon at runtime and add it to the map.
+This example uses [`addImage`](/mapbox-gl-js/api/map/#map#addimage) to dynamically generate an icon at runtime and add it to a map style. Then, it adds the icon to the map by including the `'layout': {'icon-image': ['concat', 'square-rgb-', ['get', 'color']]}` statement inside [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer).
```
nit: delete the comma after "Then"

domlet replied:
Done – thank you! ddfa1e9

ryanhamley commented on docs/pages/example/add-image-generated.md:
```diff
@@ -16,6 +16,6 @@ prependJs:
 ---
 
-Add an icon to the map that was generated at runtime.
+This example uses [`addImage`](/mapbox-gl-js/api/map/#map#addimage) to generate an icon at runtime and add it to a map style. Then, it adds the icon to the map by including the `'layout': {'icon-image': 'gradient'}` statement inside [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer).
```
nit: delete the comma after "Then"

domlet replied:
Happy to. :) c98c5a4

ryanhamley reviewed (changes requested):
This is great @domlet! I added some general comments about grammar and wording. There's only a couple technical points that I see. Let me know if you have any questions.

domlet commented on docs/pages/example/add-image-missing-generated.md:
```diff
@@ -16,6 +16,6 @@ prependJs:
 ---
 
-Dynamically generate a missing icon at runtime and add it to the map.
+This example uses [`addImage`](/mapbox-gl-js/api/map/#map#addimage) to dynamically generate an icon at runtime and add it to a map style. Then, it adds the icon to the map by including the `'layout': {'icon-image': ['concat', 'square-rgb-', ['get', 'color']]}` statement inside [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer).
```
```suggestion
This example uses [`addImage`](/mapbox-gl-js/api/map/#map#addimage) to dynamically generate an icon at runtime and add it to a map style. Then it adds the icon to the map by including the `'layout': {'icon-image': ['concat', 'square-rgb-', ['get', 'color']]}` statement inside [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer).
```

domlet commented on docs/pages/example/add-image-generated.md:
```diff
@@ -16,6 +16,6 @@ prependJs:
 ---
 
-Add an icon to the map that was generated at runtime.
+This example uses [`addImage`](/mapbox-gl-js/api/map/#map#addimage) to generate an icon at runtime and add it to a map style. Then, it adds the icon to the map by including the `'layout': {'icon-image': 'gradient'}` statement inside [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer).
```
```suggestion
This example uses [`addImage`](/mapbox-gl-js/api/map/#map#addimage) to generate an icon at runtime and add it to a map style. Then it adds the icon to the map by including the `'layout': {'icon-image': 'gradient'}` statement inside [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer).
```

domlet commented on docs/pages/example/animate-camera-around-point.md:
```diff
@@ -15,6 +15,8 @@ prependJs:
-Animate the map camera around a point.
+This example uses [`rotateTo`](/mapbox-gl-js/api/map/#map#rotateto) to animate the map camera around a point.
+
+This example uses the [`interpolate`](/mapbox-gl-js/style-spec/expressions/#interpolate) expression.
```
Done: 791cf82 – follow up question for @ryanhamley 

grammatically, when referring to expressions, should we say 

`and the interpolate expression to animate...` 

or 

`and an expression using the 'interpolate' operator to animate...`?

my sense is that expressions are the whole statements, and when we call expressions by name as in the original version here we are actually referring to the operators

would love your confirmation or clarification on this. thank you!

ryanhamley replied:
You're correct that `interpolate` is the operator, but I think for the sake of brevity it's ok to just say "expression". That's how we talk about them and I think people will know what is meant.

ryanhamley replied:
That being said, if you wanted to I think you could say "the `interpolate` expression operator" which would be fairly concise and clear.

domlet replied:
> "...and the `interpolate` expression operator..."

Wonderful, done. Thank you!

domlet replied:
25d1675

domlet commented on docs/pages/example/animate-point-along-line.md:
```diff
@@ -16,6 +16,8 @@ prependJs:
-Animate the position of a point by updating a GeoJSON source on each frame.
+This example animates the position of a point by updating a GeoJSON source on each frame.
+
+It first defines the animation as a GeoJSON object. Then it adds a [GeoJSON type source](/mapbox-gl-js/style-spec/sources/#geojson) referring to that GeoJSON, and adds a [circle type layer](/mapbox-gl-js/style-spec/layers/#circle) that uses that source. Finally, it starts the animation to continuously update the GeoJSON object, resulting in a moving circle on the map.
```
f131961

domlet commented on docs/pages/example/cluster.md:
```diff
@@ -15,6 +15,6 @@ prependJs:
 ---
 
-Use Mapbox GL JS' built-in functions to visualize points as clusters.
+This example uses Mapbox GL JS' built-in [`cluster`](/mapbox-gl-js/style-spec/sources/#geojson-cluster) functions to visualize points in a [`circle`](/mapbox-gl-js/style-spec/layers/#circle) type layer as clusters.
```
```suggestion
This example uses Mapbox GL JS' built-in [`cluster`](/mapbox-gl-js/style-spec/sources/#geojson-cluster) functions to visualize points in a [`circle`](/mapbox-gl-js/style-spec/layers/#circle) layer as clusters.
```

domlet commented on docs/pages/example/dancing-buildings.md:
```diff
@@ -15,6 +15,10 @@ prependJs:
-Connect the runtime styling API with the Web Audio API to create a map where the 3D buildings dance to the rhythm of your ambient environment.
+This example connects the runtime styling API with the Web Audio API to create a map where the 3D buildings dance to the rhythm of your ambient environment.
+
+It uses [`getUserMedia`](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia) to access audio from the user's environment, then analyzes that sound for loudness, and uses the loudness levels to control the [`fill-extrusion-height`](/mapbox-gl-js/style-spec/layers/?size=n_10_n#paint-fill-extrusion-fill-extrusion-height) property of the [`fill-extrusion`](/mapbox-gl-js/style- ...
```
removed 'type' from remaining instances of "x type source" and "x type layer" in this PR: 7c762fa 👍

domlet replied:
@ryanhamley Unresolving this conversation to share this new version of the description – everything look accurate? Since this example uses so many expressions I'd like to make sure we say something about what they are doing:

> This example uses [runtime styling](https://docs.mapbox.com/help/glossary/runtime-styling/) with the [Web Audio API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API) to create a map where the 3D buildings dynamically change height to the rhythm of your ambient environment, giving the appearance of dancing.
> 
> It uses [`getUserMedia`](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia) to access audio from the user's environment, then analyzes that sound for loudness and uses the loudness levels to control the [`fill-extrusion-height`](/mapbox-gl-js/style-spec/layers/?size=n_10_n#paint-fill-extrusion-fill-extrusion-height) property of the [`fill-extrusion`](/mapbox-gl-js/style-spec/layers/#fill-extrusion) layer.
> 
> It uses the [`==`](/mapbox-gl-js/style-spec/expressions/#==), [`>`](/mapbox-gl-js/style-spec/expressions/#%3E) and [`<=`](/mapbox-gl-js/style-spec/expressions/#%3C=) expression operators to filter buildings in the layer based on their height data.

ryanhamley replied:
"adjust" might be better than "control" in the second paragraph. For the third paragraph, maybe specify that this is done for performance reasons. Otherwise, looks good.

domlet replied:
You got it – thanks for your help demystifying this one! Updated desc:

> This example uses [runtime styling](https://docs.mapbox.com/help/glossary/runtime-styling/) with the [Web Audio API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API) to create a map where the 3D buildings dynamically change height to the rhythm of your ambient environment, giving the appearance of dancing.
> 
> It uses [`getUserMedia`](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia) to access audio from the user's environment, then analyzes that sound for loudness and uses the loudness levels to adjust the [`fill-extrusion-height`](/mapbox-gl-js/style-spec/layers/?size=n_10_n#paint-fill-extrusion-fill-extrusion-height) property of the [`fill-extrusion` layer](/mapbox-gl-js/style-spec/layers/#fill-extrusion).
> 
> To optimize performance, it uses the [`==`](/mapbox-gl-js/style-spec/expressions/#==), [`>`](/mapbox-gl-js/style-spec/expressions/#%3E) and [`<=`](/mapbox-gl-js/style-spec/expressions/#%3C=) expression operators to filter buildings in the layer based on their height data.

domlet commented on docs/pages/example/disable-rotation.md:
```diff
@@ -15,6 +15,7 @@ prependJs:
 ---
 
-Prevent users from rotating a map.
+This example uses `dragRotate.disable` and `touchZoomRotate.disableRotation` so the [`DragRotateHandler`](/mapbox-gl-js/api/handlers/#dragrotatehandler) and [`TouchZoomRotateHandler`](/mapbox-gl-js/api/handlers/#touchzoomrotatehandler) prevent a user's drag and touch interactions from rotating a map.
```
6b01193

domlet commented on docs/pages/example/drag-a-point.md:
```diff
@@ -15,6 +15,8 @@ prependJs:
-Drag the point to a new location on a map and populate its coordinates in a display.
+In this example, a user can drag a point to a new location on the map, and a map overlay displays the new point coordinates.
+
+The example stores the point coordinates in a variable with a GeoJSON value, then uses [`MapTouchEvent`s](/mapbox-gl-js/api/events/#maptouchevent) to prevent the default map drag behavior and get the updated coordinates. Then, it uses [`setData`](https://docs.mapbox.com/mapbox-gl-js/api/sources/#geojsonsource#setdata) to enable changes to the variable, and [`addSource`](https://docs.mapbox.com/ma ...
```
e7127b8

domlet commented on docs/pages/example/fullscreen.md:
```diff
@@ -15,6 +15,8 @@ prependJs:
 ---
 
-Toggle between current view and fullscreen mode.
+This example adds a clickable map control that enables a user to toggle fullscreen mode on and off.
```
I'm going to leave this one as-is because the description pattern (plain language first, technical language second) makes the example accessible to more readers.

domlet commented on docs/pages/example/hillshade.md:
```diff
@@ -17,6 +17,6 @@ prependJs:
 ---
 
-Add raster hillshading to a map.
+This example adds raster hillshading to a map by first creating a map with no hillshading, then using [`addSource`](/mapbox-gl-js/api/map/#map#addsource) and [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add the freely available [Mapbox Terrain-RGB](https://docs.mapbox.com/help/troubleshooting/access-elevation-data/#mapbox-terrain-rgb) raster tileset as a [`raster-dem`](/mapbox-gl-js/style ...
```
53f9eda

domlet commented on docs/pages/example/updating-choropleth.md:
```diff
@@ -16,6 +16,10 @@ prependJs:
-Use 2014 census data to display state or county population depending on zoom level.
+This map uses 2014 U.S. Census data to create two layers. One layer contains a choropleth visualization for _state populations_, and one layer contains a choropleth visualization for _county populations_. 
+
+The example defines a zoom level at which each layer and its respective legend should appear or disappear using the `minzoom` and `minzoom` [layer properties](/mapbox-gl-js/style-spec/layers/).
```
293a686 7ccdd19 👍

domlet commented on docs/pages/example/updating-choropleth.md:
```diff
@@ -16,6 +16,10 @@ prependJs:
+
+The example defines a zoom level at which each layer and its respective legend should appear or disappear using the `minzoom` and `minzoom` [layer properties](/mapbox-gl-js/style-spec/layers/).
+
+You can use [`getZoom`](/mapbox-gl-js/api/map/#map#getzoom) with the [`MapBoxZoomEvent`](/mapbox-gl-js/api/events/#mapboxzoomevent) event type to get zoom values to control the appearance of other elements. 
```
cf0ef1a

domlet commented on docs/pages/example/image-on-a-map.md:
```diff
@@ -16,6 +16,10 @@ prependJs:
 ---
 
-Add a radar weather image overlay on top of a dark vector baselayer.
+This example adds a radar weather image overlay on top of a dark vector baselayer.
```
thank you – I've made several changes removing instances of `baselayer` 3435885

```
description: Add a radar weather image overlay on top of a dark vector map.
[...]
This example animates a series of radar weather images over other map layers in a dark vector style.

To create the vector map style, 
```

@ryanhamley ok with you if we change the title of this example from
`Add an image`
to
`Add a raster image to a map layer`?

ryanhamley replied:
I think "to a map layer" is not strictly correct and could be a bit confusing. `Add a raster image to a map` should be good though

domlet commented on docs/pages/example/multiple-geometries.md:
```diff
@@ -15,6 +15,8 @@ prependJs:
-Add a polygon and circle layer from the same GeoJSON source.
+This example uses [`addSource`](/mapbox-gl-js/api/map/#map#addsource) to add one GeoJSON source to a map. The source contains both polygon and point features.
+
+Then, it uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add one `fill` layer, which will render the polygons, and one `circle` layer, which will render the points, to the map.
```
```suggestion
Then it uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add one `fill` layer, which will render the polygons, and one `circle` layer, which will render the points, to the map.
```

domlet commented on docs/pages/example/query-similar-features.md:
```diff
@@ -15,6 +15,10 @@ prependJs:
 ---
 
-Hover over counties to highlight counties that share the same name.
+In this example, a user can hover their cursor over any county in the United States, and the map highlights other U.S. counties with a matching name. The application also shows a popup with the name of the county at the cursor location, and shows an HTML overlay with summary information.
```
Leaving first comma as-is because I think it improves readability for readers with English not as a first language (`the map highlights other U.S. counties with a matching name` is not a thing that `a user can hover their cursor over`).

domlet commented on docs/pages/example/simple-map.md:
```diff
@@ -16,6 +16,8 @@ prependJs:
-Initialize a map in an HTML element with Mapbox GL JS.
+This example uses [`mapboxgl.Map`](/mapbox-gl-js/api/map) to initialize a Mapbox map inside an HTML element on a webpage.
+
+You can use [`map` properties](/mapbox-gl-js/api/map/#map-parameters) like `style`, `center`, and `zoom` in the declaration to define the initial appearance of the map.
```
👍 c11f558

domlet commented on docs/pages/example/toggle-layers.md:
```diff
@@ -15,6 +15,8 @@ prependJs:
-Create a custom layer switcher to display different datasets.
+This example adds a clickable interface that allows a user to enable and disable two different map layers.
+
+The interface uses [`setLayoutProperty`](/mapbox-gl-js/api/map/#map#setlayoutproperty) to toggle the value for each layer's [`layout`](/mapbox-gl-js/style-spec/layers/#layout) sub-property for `visibility` between `visble` and `none`.
```
Thank you! 2 of 3 changes made: 32b1148

> What about "to toggle the value for each layer's visibility property between visible and none" and link directly to the visibility property rather than to layout?

^ We must link to `layout` because `visibility` is a sub-property that appears 10 times on the doc for `layout` – once for each layer type, with differing versions for SDK support across layer types.

domlet commented on docs/pages/example/video-on-a-map.md:
```diff
@@ -16,6 +16,8 @@ prependJs:
 ---
 
-Display a video on top of a satellite raster baselayer. Click the map to play and pause the video.
+Display a [`video`](/mapbox-gl-js/style-spec/sources/#video) in a map layer by adding a [source](/mapbox-gl-js/style-spec/sources/) and then adding a [layer](/mapbox-gl-js/style-spec/layers/) using that source. This example adds a video layer on top of a raster layer with satellite imagery.
```
It's true that this pattern should be familiar to users familiar with GL JS, but docs team assumes that any example could be a user's first exposure to GL JS. 

I've made some improvements to the description though: 5b2e5087dbb4b21cfe7dd7c861566b62c4ad9d8b

domlet commented on docs/pages/example/query-similar-features.md:
```diff
@@ -15,6 +15,10 @@ prependJs:
+
+The example first uses [`addSource`](/mapbox-gl-js/api/map/#map#addsource) to connect data from a vector tileset, and then uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add two [`fill`](/mapbox-gl-js/style-spec/layers/#fill) layers using that source.
+
+It uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to find the name of the county located at the cursor position, then uses [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to find other counties with the same name and highlights them using an overlay on the map.
```
thank you! I've created https://github.com/mapbox/mapbox-gl-js-docs/issues/478 to track updating the example and here is the revised description – look ok?

> In this example, a user can hover their cursor over any county in the United States, and the map highlights other U.S. counties with a matching name, displays a popup with the name of the county at the cursor location, and displays an HTML overlay with summary information.
> 
> The example first uses [`addSource`](/mapbox-gl-js/api/map/#map#addsource) to connect data from a vector [tileset](https://docs.mapbox.com/help/glossary/tileset/). Then it uses [`addLayer`](/mapbox-gl-js/api/map/#map#addlayer) to add two [`fill` layers](/mapbox-gl-js/style-spec/layers/#fill) using that source, one for the a default display and one with all features highlighted with a color.
> 
> To add interactivity, it first uses [`mousemove`](/mapbox-gl-js/api/map/#map.event:mousemove) to find the name of the county located at the cursor position, then uses [`querySourceFeatures`](/mapbox-gl-js/api/map/#map#querysourcefeatures) to create a filter including only counties with a matching name, and finally uses [`setFilter`](/mapbox-gl-js/api/map/#map#setfilter) to add only matching features to the highlight layer.

ryanhamley replied:
>The example first uses addSource to connect data from a vector tileset. Then it uses addLayer to add two fill layers using that source, one for the a default display and one with all features highlighted with a color.

I think "connect" isn't the right choice here; "connect" the data to what? I would just say that it uses `addSource` to add the county polygons as a vector tileset.

The way this works is that we're adding two layers, one with transparent county polygons (the `rgba(0,0,0,0.1)` part means that the color is black at 10% opacity for the fill color and fill outline color) and one with filled county polygons. The filled county polygon layer has a filter on it that makes all the features in this layer not visible by default (it filters for the county name being an empty string which doesn't match anything). On `mousemove`, we use the mouse event to get the county feature that the mouse is currently over (`var feature = e.features[0];`). We use `querySourceFeatures` to get all the county features that match the current county name. We use that list of features to get the total population and create the overlay which is actually the white HTML box that's added to the map. Then we update the filter that's applied to the `counties-highlighted` layer (so instead of `['in', 'COUNTY', '']` it is now `['in' 'COUNTY', 'Washington']` for example). This then allows the matching counties to be visible, creating the illusion of a highlight feature. Finally, on `mouseleave`, we reset the `counties-highlighted` filter to the empty string and remove the overlay box.

domlet replied:
Thank you! I've made some improvements to both the code comments and the example description – I think we are good: c11fd627946d259e355a2f4d8adf6c98efd5ac08

domlet reviewed (commented):
I'm not sure how I made myself a reviewer on my own PR, but I'm approving here so my replies to Ryan do not disappear.

ryanhamley reviewed (approved):
I think this looks good from my standpoint. There's one thing I'd still change which is the `query-similar-features.md` description. See my response [here](https://github.com/mapbox/mapbox-gl-js-docs/pull/474#discussion_r560618484) for more detail.

domlet commented:
Thanks so much @HeyStenson and @ryanhamley for your detailed reviews and responses.

@ryanhamley If you want final eyes on these two, here are the two last descriptions reformulated with both of your feedback:

1. generate missing icon - https://github.com/mapbox/mapbox-gl-js-docs/pull/474#discussion_r561360178
2. query similar features - https://github.com/mapbox/mapbox-gl-js-docs/commit/c11fd627946d259e355a2f4d8adf6c98efd5ac08#diff-469672edfba34847d1ca0eb5565faeb5733cbe8f453873741f488863ffcb7cb5

@HeyStenson I'll leave this open for your final review. Thank you!

ryanhamley commented:
Looks great @domlet!

HeyStenson commented:
fixes https://github.com/mapbox/mapbox-gl-js-docs/issues/432
fixes https://github.com/mapbox/mapbox-gl-js-docs/issues/409
fixes https://github.com/mapbox/mapbox-gl-js-docs/issues/351
fixes https://github.com/mapbox/mapbox-gl-js-docs/issues/290

HeyStenson reviewed (approved):
🍊 Amazing lift, @domlet!

domlet commented:
Thank you both! And thank you @HeyStenson for noting the [tickets we can close](https://github.com/mapbox/mapbox-gl-js-docs/pull/474#issuecomment-764168189)!
